pkg/server/filters/metrics_server.go

```diff
@@ -0,0 +1,144 @@
+package filters
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/loft-sh/vcluster/pkg/server/handler"
+	requestpkg "github.com/loft-sh/vcluster/pkg/util/request"
+	"github.com/loft-sh/vcluster/pkg/util/translate"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog"
+	"k8s.io/metrics/pkg/apis/metrics"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	NodeResource = "nodes"
+	PodResource  = "pods"
+	APIVersion   = "v1beta1"
+
+	HeaderContentType = "Content-Type"
+)
+
+func WithMetricsServerProxy(h http.Handler, cacheHostClient, cachedVirtualClient client.Client, hostConfig *rest.Config) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		info, ok := request.RequestInfoFrom(req.Context())
+		if !ok {
+			requestpkg.FailWithStatus(w, req, http.StatusInternalServerError, fmt.Errorf("request info is missing"))
+			return
+		}
+
+		if isMetricsServerProxyRequest(info) {
+			splitted := strings.Split(req.URL.Path, "/")
+
+			metricsServerProxy := &MetricsServerProxy{
+				codecFactory:   serializer.NewCodecFactory(cachedVirtualClient.Scheme()),
+				request:        req,
+				requestInfo:    info,
+				responseWriter: w,
+				resourceType:   NodeResource,
+			}
+
+			if info.Resource == PodResource {
+				namespace := translate.Default.PhysicalNamespace(info.Namespace)
+				name := translate.Default.PhysicalName(info.Name, namespace)
+
+				metricsServerProxy.resourceType = PodResource
+
+				// replace the translated name and namespace
+				splitted[5] = namespace
+				splitted[7] = name
```

Comment:
Would be good to check the size before accessing the array to avoid unexpected panic.
e.g. when the List request is done, is there anything in the seventh position?

Reply:
I think this is the cause for panic triggered by the `kubectl get pods.metrics.k8s.io -A`

Reply:
Yea I agree and I was just thinking instead of only relying on `len()` I wanted to use a more concrete thing like info from `requestinfo`, I have handled that case now